### ISSUE TITLE
[BUG]  Prism.Container.Abstractions, Version=9.0.107.57918, Culture=neutral, PublicKeyToken=40ee6c3a2184dc59  Or one of its dependencies. Strong name verification failed

### ISSUE THREAD

**Issue by weiwei11245** [PrismLibrary/Prism#3235]
### Description

When I upgraded from Prism 8 to 9, this error occurred  

An unprocessed exception of type 'System. IO. FileLoadException' is in an unknown module. Occurred in the middle
Failed to load file or assembly 'Prism. Container. Abstracts, Version=9.0.107.57918, Culture=neutral, PublicKeyToken=40ee6c3a2184dc59' or one of its dependencies. Strong name validation failed. (Exception from HRESULT: 0x8013141A)

### Steps to Reproduce

What happens during program startup 

### Platform with bug

WPF

### Affected platforms

Windows

**Comment by volvo245r** [comment]
I have exactly the same issue, 
Exception thrown: 'System.IO.FileLoadException' in PresentationFramework.dll
An unhandled exception of type 'System.IO.FileLoadException' occurred in PresentationFramework.dll
Could not load file or assembly 'Prism.Container.Abstractions, Version=9.0.107.57918, Culture=neutral, PublicKeyToken=40ee6c3a2184dc59' or one of its dependencies. Strong name validation failed. (Exception from HRESULT: 0x8013141A)

bin\Release>sn -v Prism.Container.Abstractions.dll

Microsoft (R) .NET Framework Strong Name Utility  Version 4.0.30319.0
Copyright (c) Microsoft Corporation.  All rights reserved.

Prism.Container.Abstractions.dll is a delay-signed or test-signed assembly

**Comment by qtbui159** [comment]
same error

**Comment by dansiegel** [comment]
Yes we're aware of the issue though I'm not sure why this broke. We've been using the same logic to strong name the assemblies since the beginning of Prism.

**Comment by letscodewithkalyan** [comment]
Error (10255) / AndroidRuntime: android.runtime.JavaProxyThrowable: [System.TypeLoadException]: Could not resolve type with token 01000061 from typeref (expected class 'System.ExceptionHelper' in assembly 'Prism.Container.Abstractions, Version=9.0.106.9543, Culture=neutral, PublicKeyToken=40ee6c3a2184dc59')

Similar error with the 
Prism.DryIoc.Maui package with latest Version=9.0.537 In Android

**Comment by tkcecek** [comment]
registry trick from https://stackoverflow.com/questions/403731/strong-name-validation-failed worked for me

**Comment by Mr-Cat-r** [comment]
As of today, this BUG still exists. I have no choice but to roll back to version 8.1.97, and then it can run normally. I have tried various methods.  【强名称验证失败】

**Comment by dansiegel** [comment]
Correct it still exists. And we're working on some back channels to fix this with Microsoft because there is no reason for this to have broken. We haven't changed anything.

**Comment by smsaleem** [comment]
When a [signed] assembly is deployed using ClickOnce, its updates are also hash-checked by ClickOnce security.  If the assembly uses different signing certificate in its subsequent version, although it compiles successfully, assembly signing error is raised by ClickOnce when updates are deployed on client machine.

Please check if it helps.

**Comment by AbraxasCSharp** [comment]
Same problem here, we have to downgrade to 8.1 too.

**Comment by nvonada** [comment]
We have two Prism exes building in the same solution.  Once we updated to 9 one had this problem and the other did not.  It turns out that while both had a platform target of "Any CPU" the one that _failed_ had the "Prefer 32-bit" checkbox selected.  I cleared that checkbox and the problem went away.  YMMV!

**Comment by toumir** [comment]
I'm still encountering the error described in the issue, any additional advice or workarounds would be greatly appreciated

**Comment by dansiegel** [comment]
Unfortunately I haven't been able to resolve this with Microsoft yet so there isn't an update to give. All I can tell you is that the process we use for the strong naming hasn't changed in years so we're not sure why the tooling isn't doing what it's supposed to.

**Comment by Bodewes** [comment]
I've the same issue after updating from 8.1.97 to 9.0.537.

Sdk="Microsoft.NET.Sdk"
TargetFramework=net472
Win11
VisualStudio 17.10.3

**Comment by nvonada** [comment]
> We have two Prism exes building in the same solution. Once we updated to 9 one had this problem and the other did not. It turns out that while both had a platform target of "Any CPU" the one that _failed_ had the "Prefer 32-bit" checkbox selected. I cleared that checkbox and the problem went away. YMMV!
> 
> Addendum: Once the apps built and ran (no detected problems BTW!) I ran unit tests. Three unit test assemblies also threw strong name errors, all were compiled x86. Switching to AnyCPU fixed them. Something 32/64 bit going on here?

OK, I thought I had this beat but my PR build is blowing up on Azure DevOps.  2/3 of my unit tests will not run on the build server due to this error.  I am going to try a couple of things but I might be defeated here.

**Comment by brianlagunas** [comment]
Is everyone here reporting this issue using WPF?

What if we just removed strong name signing. Strong-names are left over from previous eras of .NET where needed to differentiate between code that was trusted, versus code that was untrusted.

Strong naming is about identity not security.

Thoughts?

**Comment by AbraxasCSharp** [comment]
Yes I'm reporting it for WPF. Removing the strong name signing would help a lot.
I can only support that Strong naming is about identity.

**Comment by smsaleem** [comment]
ClickOnce needs strong naming.  Is there anyway to bypass it ?

**Comment by brianlagunas** [comment]
You can temporarily disable strong name validation for the delay-signed assembly using the `sn.exe` tool.

```
sn -Vr Prism.Container.Abstractions,40ee6c3a2184dc59
```

After disabling strong name validation, rebuild your project and run it again. This should bypass the strong name validation failure and allow the assembly to load properly.

**Comment by Bodewes** [comment]
> You can temporarily disable strong name validation for the delay-signed assembly using the `sn.exe` tool.
> 
> ```
> sn -Vr Prism.Container.Abstractions,40ee6c3a2184dc59
> ```
> 
> After disabling strong name validation, rebuild your project and run it again. This should bypass the strong name validation failure and allow the assembly to load properly.

If I go this route, I have to do this on every PC runnning the WPF application? I don't think Corparate(tm) will like that.

**Comment by toumir** [comment]
Hi @brianlagunas,

Thanks for the tip on disabling strong name validation. **Unfortunately**, I'm still encountering the System.IO.FileLoadException even after turning it off.

The issue seems to be related to using the latest versions of the Prism packages:

Prism.Container.Abstractions (version: 9.0.107)
Prism.Container.DryIoc (version: 9.0.107)
Prism.Core (version: 9.0.537)
Prism.DryIoc (version: 9.0.537)
Prism.Events (version: 9.0.537)
Prism.Wpf (version: 9.0.537)
When running the application, we get a **strange** exception indicating a dependency on an older version (9.0.106.9543) of Prism.Container.Abstractions.

Here's the exact error message:

System.IO.FileLoadException: 'Could not load file or assembly 'Prism.Container.Abstractions, Version=**9.0.106.9543**, Culture=neutral, PublicKeyToken=40ee6c3a2184dc59' or one of its dependencies. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)'   

Any ideas on how to resolve this version mismatch?

**Comment by chk040399** [comment]
Maybe if we have access to the Source code for these packages (Prism.Container.Abstractions and Prism.Container.DryIoc), we can build unsigned versions of these assemblies, as the source repository for these assemblies is not available (https://github.com/PrismLibrary/Prism.Containers).

**Comment by dansiegel** [comment]
@chk040399 I can sympathize with your frustration. The reality is that we actually haven't changed anything about the way that we handle package signing and strong naming, and it wasn't until we finally released the GA that this was reported. This should highlight to the community the importance of using preview packages so that we can catch issues like this sooner. 

Something in the build system has caused the problem and we are still looking into this. While I cannot tell you exactly when we will be able to resolve it, we are continuing to look into this.

**Comment by Shannon-roos** [comment]
I can also report haivng this after updating to 9.0.107.

The registry trick from @DevMercenary has seemingly done the trick for development, however like @Bodewes this isn't a great option for production. 

If you need anything from me to test\replicate\verify please let me know and I can do whatever needed.

**Comment by maho123456emer** [comment]
Same issue here using WPF. I only have this issue with version 9.0.537. With 9.0.401-pre everything works fine.

**Comment by AnilNalawade** [comment]
> Prism.Container.Abstractions (version: 9.0.107)

Same error - 'System.IO.FileLoadException' I am getting for Prism.Container.Abstractions (version: 9.0.107).

**Comment by bart-fan** [comment]
Mark it and wait for the problem to be solved

**Comment by bart-fan** [comment]
> Mark it and wait for the problem to be solved

Now .use 8.1.97

**Comment by dspear** [comment]
Has there been a resolution to this problem? I have gone through a lot of effort to update from a very old Prism to the latest, but this problem is starting to look like a showstopper for the whole update project.

**Comment by brianlagunas** [comment]
@dspear You can use the preview version 9.0.401-pre which was the release before the current version.

**Comment by brianlagunas** [comment]
Can everyone confirm this only happens in .NET Framework, and not .NET Core?  I am unable to replicate this for .NET Core, but I can get it to happen for .NET Framework. Specifically, .NET Framework 4.7.2.

**Comment by Shannon-roos** [comment]
@brianlagunas 

I'm seeing it on .NET Framework 4.8 and did not see it on Prism V 8.1.97 that I updated from. 

Something of note is that I'm not seeing everywhere I've installed. I can try and get the system info of some that have worked\not worked if that adds anything useful? Anecdotally I've encountered more Not OKs then OKs. 

Initial Examples off-hand: 

**Not OK (Until @DevMercenary fix):**
Windows 11 Home (24HS) (Clean install) 

**OK:**
Windows 11 Pro for Workstations (23H2) (Existing install) 
Windows Server 2012 R2 (Old I know)

**Comment by maho123456emer** [comment]
I can confirm the comment from @Shannon-roos. I also have this problem when using .NET Framework 4.8 in combination with version 9.0.537.
When using .NET Framework 4.8 in combination with 9.0.401-pre, everything works fine.

**Comment by nvonada** [comment]
We saw it in .NET Framework 4.8.

**Comment by dspear** [comment]
We see it with .Net Framework 4.8, and 9.0.401-pre works fine.

I expect that the machines that it works on already have the registry "fix" in place. We found some machines that had a "*,*" key under HKLM\SW\MS\StrongName\Verification. With that in place, the StrongName checking is not done.

**Comment by brianlagunas** [comment]
This appears to be related to our GitHub actions. Even our own unit test projects fail to load the .NET Framework assembly with the same error message.

![Image](https://github.com/user-attachments/assets/bf6aaf9b-3a4f-49a3-bea9-b648e297f40b)

For the life of me I can't figure out why it fails for .NET Framework and works for .NET Core

This is a head scratcher.

**Comment by maho123456emer** [comment]
Hello @brianlagunas,
what do you mean by “everything works locally”? We can reproduce this bug without GitHub. When we build our application locally with Visual Studio, we get the error as well. So for us it doesn't only occur when we build our application on a build server.

Do you need a small sample application where the error occurs (locally)?

**Comment by brianlagunas** [comment]
@maho123456emer I am talking about building and signing the assemblies locally. I have access to the containers source code. So it appears on the surface that the signing error is occurring in our build process in our CI/CD pipeline.

**Comment by marksmeltzer** [comment]
> For the life of me I can't figure out why it fails for .NET Framework and works for .NET Core.

AFAIK, .NET Core doesn't use strong names anymore. Outmoded concept.

See this comment from [jkotas](https://github.com/jkotas):

https://github.com/dotnet/docs/issues/22412#issuecomment-764395520

And in the official doc from 2021:
https://learn.microsoft.com/en-us/dotnet/standard/assembly/strong-named

> For .NET Core and .NET 5+, strong-named assemblies do not provide material benefits. **The runtime never validates the strong-name signature, nor does it use the strong-name for assembly binding.**

**Comment by nomadev** [comment]
While investigating the signing issue, I ran the following command under the NuGet folder in the user profile:  
```powershell
.\sn.exe -v "Prism.Container.Abstractions.dll"
```  
If the assembly is correctly strong-name signed, the expected output is:  
`*assemblyName* is valid`  

However, the result was:  
`Prism.Container.Abstractions.dll is a delay-signed or test-signed assembly`

Notably, the application works if strong name verification is disabled for this assembly using:  
```powershell
.\sn.exe -Vr *,public key token
```

**Comment by brianlagunas** [comment]
@nomadev Thanks for sharing. That is extremely strange since we aren't delaying signing anything. Something very strange is going on here.

**Comment by DmitryLukyanov** [comment]
>.\sn.exe -Vr *,public key token

this approach doesn't work for me, even though I see this output:

        PS C:\Program Files (x86)\Microsoft SDKs\Windows\v7.0A\bin> .\sn.exe -Vr "C:\Git\WpfAppDemo\WpfAppDemo\bin\Debug\Prism.Container.Abstractions,40ee6c3a2184dc59"
        
        Microsoft (R) .NET Framework Strong Name Utility  Version 3.5.30729.1
        Copyright (c) Microsoft Corporation.  All rights reserved.
        
        Verification entry added for assembly 'C:\Git\WpfAppDemo\WpfAppDemo\bin\Debug\Prism.Container.Abstractions,40ee6c3a2184dc59'

when I rebuild and start an application. I still see this error:

      An unhandled exception of type 'System.IO.FileLoadException' occurred in Unknown Module.
      Could not load file or assembly 'Prism.Container.Abstractions, Version=9.0.106.9543, Culture=neutral, PublicKeyToken=40ee6c3a2184dc59' or one of its dependencies. Strong name validation failed. (Exception from HRESULT: 0x8013141A)

same error happens for `Prism.Container.Abstractions` with latest version `9.0.107`

**Comment by maho123456emer** [comment]
@brianlagunas We have just purchased a commercial licence for Prism, but we are currently unable to use it. We are forced to use the MIT licensed 8.x version because of this bug. This bug has been open for half a year now and seems to be difficult to fix, all good. But to be honest, is there any indication of when we can expect this problem to be fixed? It would be nice if this bug could be treated with higher priority.

**Comment by maho123456emer** [comment]
Yeah the 9.0.401-pre version works, but we want to wait for this bug to be fixed before we update to 9.x.
It is good to know that you are still working on this bug. Hopefully you will be successful soon 🚀

Thanks!

**Comment by brianlagunas** [comment]
Okay, I have made some progress. I have narrowed down the issue to something going on between our github actions. I found the point where the assemblies are actually being signed. I am going to continue painstakingly working through the action to find out the root cause. But I have made positive progress :)

**Comment by brianlagunas** [comment]
This has been fixed. Commercial Plus license holders can get the latest build with `v9.1.65-pre`.

For those Commercial Plus customers who would like a backport to `v9.0.x` let me know and I'll see what I can do about that.

**Comment by brianlagunas** [comment]
I went ahead and backported the fix to `v9.0.540`. Commercial Plus customers have access to this backport now.

**Comment by nvonada** [comment]
Yes! Thank you ever so much.  This must have been a nightmare to track down.

**Comment by ZGGSONG** [comment]
My WPF application requires the .NET Framework 4.7.2, and directly upgrading to Prism 9.x results in errors. If I downgrade the WPF program's .NET version to 8.0, the issue disappears.

**Comment by MinghuaOrcl** [comment]
Do you have plan to update Nuget version to 9.0.540?

**Comment by brianlagunas** [comment]
No, the community version does not get backported fixes. That is reserved for our commercial plus holders only.

**Comment by maho123456emer** [comment]
When will there be a new version for **Commercial** license holders with this bug fix?

**Comment by brianlagunas** [comment]
Commercial licenses holders do not get access to the private nuget server and must rely on the public nuget.org packages which will always be behind. We currently don't have a timeline for when the public nuget will be updated.

**Comment by brianlagunas** [comment]
This issue has already been fixed. Only Commercial Plus licenses holders have access to the latest nuget packages via their private feed.
